dulwich/porcelain.py

```diff
@@ -1753,6 +1755,76 @@ def update_head(repo, target, detached=False, new_branch=None):
             r.refs.set_symbolic_ref(b"HEAD", to_set)
 
 
+def reset_file(repo, file_path: str , target: bytes = b'HEAD'):
+    """
+    reset the file to specific commit or branch
+    Args:
+        repo: dulwich Repo object
+        file_path: file to reset, relative to the repository path
+        target: branch or commit or b'HEAD' to reset
+    """
+    if target in branch_list(repo):
+        sha = repo.refs[b'refs/heads/' + target]
+    elif target == b'HEAD':
+        sha = repo.head()
+    else:
+        sha = target
+    tree = parse_tree(repo, treeish=sha)
+
+    def get_entry(repo, tree, file_path: str):
+        if os.path.split(file_path)[0] != '':  # file_path with directory
+            par_dir_tree_id = get_entry(repo, tree, os.path.split(file_path)[0])[1]
+            tree = parse_tree(repo, treeish=par_dir_tree_id)
+            tree_id = tree[os.path.split(file_path)[1].encode()]
+            return tree_id
+        else:
+            try:
+                return tree[file_path.encode()]
+            except KeyError:
+                # untracked file not in tree.
+                raise KeyError("file '%s' not in tree" % (file_path))
+
+    file_entry = get_entry(repo, tree, file_path)
+    
+    full_path = os.path.join(repo.path, file_path)
+    blob = repo.object_store[file_entry[1]]
+    mode = file_entry[0]
+    build_file_from_blob(blob, mode, full_path.encode())
+
+
+def checkout(repo, branch: bytes, force: bool = False):
+    """
+    switch branches or restore working tree files
+    Args:
+        repo: dulwich Repo object
+        branch: branch name to checkout
+    """
+
+    # check repo status
+    if not force:
+        if status(repo) != GitStatus(staged={'add': [], 'delete': [], 'modify': []}, unstaged=[], untracked=[]):
+            raise CheckoutError('working directory not clean')
+    update_head(repo, branch)
+
+    # unstage all files in the index
+    tracked_changes = []
+    for file in get_tree_changes(repo).values():
+        tracked_changes += file
+    repo.unstage(tracked_changes)
+
+    # reset all unstaged file to target
+    normalizer = repo.get_blob_normalizer()
+    filter_callback = normalizer.checkin_normalize
+    unstaged_files = list(get_unstaged_changes(repo.open_index(), repo.path, filter_callback))
+    for file in unstaged_files:
+        reset_file(repo, file.decode(), b'HEAD')
```

Comment:
wouldn't this need to update the index as well?

Reply:
reset_file will update index now

Reply:
Where we should update index?After reset_file or at the end of checkout?

Reply:
After the checkout, otherwise this will be very slow for larger repositories.

Reply:
Have  I update index in the stage()?Or I should update index on more time?

Reply:
So ideally we'd just update the index once for each checkout operation, since it can be expensive for larger indexes. I'm okay with compromising for now and e.g. doing one write for staging and one for unstaging, but one write for each will be noticeable.

Reply:
sorry, I can’t understand you.Since we have update index in the Repo.unstage, should we update index in the end of checkout?

Reply:
Yeah, we should update the index at the end of checkout - but we can't be updating the index for every reset_file() call, since that simply won't scale.

Reply:
Repo.unstage will update the different file's index between tatget branch and current branch, should we update the index at the end of checkout.

Reply:
Yes -  we shouldn't be modifying the index every time reset_file() gets called. The changes *after* unstage still need to be properly written to the index.

Reply:
So I think we just update index after unstage and needn't update index at the end of checkout.

Reply:
You'd have to pass index to reset_file() in that case though, and you may want to make it private - or provide a _reset_file() that works on an index and is used by both checkout() and a public reset_file() wrapper that takes a repo argument.

Reply:
why should I pass index to reset_file()? I needn't index in reseet_file().